Web panics after upgrade to 5.7.2

## Bug Report

We are running concourse for around 60 teams on k8s. Each teams gets it's own concourse instance with two web pods backed by RDS and 1-n autoscaling ephemeral workers pods.
For one team we saw the k8s health checks on /api/v1/info failing on the web pods with a connection refused.
After digging into this we could not find any underlying infrastructure problems and all other teams worked fine.

At the time of the failure, the web pod's logs showed a nil pointer dereference runtime panic.

The error first occurred 10th of December. We upgraded the concourse version to 5.7.2 on 30th of November. Not sure if this is related

Logs (newest on top)
```
2019-12-10 17:40:44	
{"timestamp":"2019-12-10T16:40:44.833455239Z","level":"info","source":"atc","message":"atc.cmd.start","data":{"session":"1"}}
2019-12-10 17:40:43	
/tmp/build/1c3187db/concourse/atc/radar/runner.go:113 +0x316
2019-12-10 17:40:43	
created by github.com/concourse/concourse/atc/radar.(*Runner).scanResources
2019-12-10 17:40:43	
/tmp/build/1c3187db/concourse/atc/radar/runner.go:118 +0x147
2019-12-10 17:40:43	
github.com/concourse/concourse/atc/radar.(*Runner).scanResources.func1(0xc001496050, 0x28a0780, 0xc0008c5c20, 0x2883b80, 0xc001490240, 0x28cf260, 0xc00011c6e0, 0xc001925080, 0x13)
2019-12-10 17:40:43	
/tmp/build/1c3187db/concourse/atc/radar/interval_runner.go:68 +0x2fa
2019-12-10 17:40:43	
github.com/concourse/concourse/atc/radar.(*intervalRunner).Run(0xc0018f0370, 0x2883b80, 0xc001490240, 0x0, 0x0)
2019-12-10 17:40:43	
/tmp/build/1c3187db/concourse/atc/radar/resource_scanner.go:64 +0x97
2019-12-10 17:40:43	
github.com/concourse/concourse/atc/radar.(*resourceScanner).Run(0xc001502090, 0x28a0780, 0xc001412180, 0x3d, 0xdf8475801, 0x0, 0x0)
2019-12-10 17:40:43	
/tmp/build/1c3187db/concourse/atc/radar/resource_scanner.go:255 +0x7d8
2019-12-10 17:40:43	
github.com/concourse/concourse/atc/radar.(*resourceScanner).scan(0xc001502090, 0x28a0780, 0xc002cc5140, 0x3d, 0x0, 0x0, 0x0, 0x0, 0x0)
2019-12-10 17:40:43	
/tmp/build/1c3187db/concourse/atc/radar/resource_scanner.go:346 +0x8f7
2019-12-10 17:40:43	
github.com/concourse/concourse/atc/radar.(*resourceScanner).check(0xc001502090, 0x28a0780, 0xc002cc5140, 0x28cf260, 0xc004bfbce0, 0x28adaa0, 0xc0031d9c80, 0xc0040bf4a0, 0xc00414f680, 0x2, ...)
2019-12-10 17:40:43	
/tmp/build/1c3187db/concourse/atc/worker/worker.go:217 +0x10c8
2019-12-10 17:40:43	
github.com/concourse/concourse/atc/worker.(*gardenWorker).FindOrCreateContainer(0xc000f05040, 0x2883bc0, 0xc000062018, 0x28a0780, 0xc002cc5140, 0x287aac0, 0x39379f8, 0x2854040, 0xc002f40cc0, 0x1f27187, ...)
2019-12-10 17:40:43	
goroutine 448 [running]:
2019-12-10 17:40:43	
2019-12-10 17:40:43	
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0xab3ec8]
2019-12-10 17:40:43	
panic: runtime error: invalid memory address or nil pointer dereference
2019-12-10 17:40:17	
{"timestamp":"2019-12-10T16:40:17.610731167Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"services","resource":"orders-src","session":"17.3.20.1.1333","team":"t***r"}}
```

## Steps to Reproduce

Unknown

## Expected Results

No runtime panic

## Actual Results

Panic happens followed by being unresponsive via http

## Additional Context

It seems like the pq: deadlock error is often preceding the panic, but not always. Might be coincidence.
Below are the logs (newest on top) over the last 30 days filtered by only these two errors.

### Filtered occurrences in Pod 1
```	
2019-12-19 09:15:23	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-19 09:15:23	
{"timestamp":"2019-12-19T08:15:23.265138002Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"services","resource":"users-artifact","session":"17.3.26.1.1","team":"t***r"}} 
2019-12-19 09:15:23	
{"timestamp":"2019-12-19T08:15:23.098706641Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"services","resource":"requests-src","session":"17.3.23.1.1","team":"t***r"}} 
2019-12-19 09:10:07	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-19 09:10:01	
{"timestamp":"2019-12-19T08:10:01.311552757Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"services","resource":"build-output-dealers","session":"17.3.4.1.1473","team":"t***r"}} 
2019-12-18 07:16:41	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-18 07:16:36	
{"timestamp":"2019-12-18T06:16:36.446956171Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"t***r","resource":"docker-cypress","session":"17.1.14.1.1","team":"t***r"}} 
2019-12-18 07:12:18	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-12 19:40:23	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-10 17:25:40	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-10 17:25:02	
{"timestamp":"2019-12-10T16:25:02.397976452Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"t***r","resource":"cf-acceptance-user","session":"17.1.9.1.1369","team":"t***r"}} 
2019-12-10 17:24:58	
{"timestamp":"2019-12-10T16:24:58.794175089Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"t***r","resource":"cf-sandbox-user","session":"17.1.12.1.1365","team":"t***r"}} 
```


###  Filtered occurrences in Pod 2

```	
2019-12-19 09:13:42	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-19 09:13:38	
{"timestamp":"2019-12-19T08:13:38.819944490Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"t***r","resource":"cf-production-user","session":"17.1.11.1.1","team":"t***r"}} 
2019-12-19 09:10:01	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-19 09:09:49	
{"timestamp":"2019-12-19T08:09:49.050526264Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"services","resource":"build-output-requests","session":"17.3.7.1.1476","team":"t***r"}} 
2019-12-18 07:30:47	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-18 07:29:04	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-18 07:28:58	
{"timestamp":"2019-12-18T06:28:58.504477540Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"t***r","resource":"e2e-src","session":"17.1.15.1.1","team":"t***r"}} 
2019-12-18 07:23:32	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-18 07:12:22	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-18 02:10:03	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-13 21:08:27	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-13 21:08:26	
{"timestamp":"2019-12-13T20:08:26.245215642Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"services","resource":"d***t","session":"17.3.46.1.2","team":"t***r"}} 
2019-12-12 19:40:11	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-11 19:22:08	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-11 19:21:33	
{"timestamp":"2019-12-11T18:21:33.736066964Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource-type.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"services","resource-type":"cf-deploy-credentials","session":"17.3.2.1.1432","team":"t***r"}} 
2019-12-10 18:14:08	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-10 18:14:04	
{"timestamp":"2019-12-10T17:14:04.051365083Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource-type.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"t***r","resource-type":"cf-cli","session":"17.1.28.1.1","team":"t***r"}} 
2019-12-10 17:40:43	
panic: runtime error: invalid memory address or nil pointer dereference 
2019-12-10 17:40:17	
{"timestamp":"2019-12-10T16:40:17.610731167Z","level":"error","source":"atc","message":"atc.pipelines.radar.scan-resource.interval-runner.tick.failed-to-create-container-in-db","data":{"error":"pq: deadlock detected","pipeline":"services","resource":"orders-src","session":"17.3.20.1.1333","team":"t***r"}} 
```



## Version Info

* Concourse version: 5.7.2
* Deployment type (BOSH/Docker/binary): Docker
* Infrastructure/IaaS: AWS
* Did this used to work?
Until 10th of December it worked

--- a/atc/worker/worker.go
+++ b/atc/worker/worker.go
@@ -201,7 +201,6 @@ func (worker *gardenWorker) FindOrCreateContainer(
 	} else if createdContainer != nil {
 		containerHandle = createdContainer.Handle()
 	} else {
-
 		logger.Debug("creating-container-in-db")
 		creatingContainer, err = worker.dbWorker.CreateContainer(
 			owner,
@@ -212,6 +211,8 @@ func (worker *gardenWorker) FindOrCreateContainer(
 			if _, ok := err.(db.ContainerOwnerDisappearedError); ok {
 				return nil, ResourceConfigCheckSessionExpiredError
 			}
+
+			return nil, err
 		}
 		logger.Debug("created-creating-container-in-db")
 		containerHandle = creatingContainer.Handle()

--- a/atc/worker/worker_test.go
+++ b/atc/worker/worker_test.go
@@ -259,7 +259,6 @@ var _ = Describe("Worker", func() {
 		fakeDBWorker.HTTPProxyURLReturns("http://proxy.com")
 		fakeDBWorker.HTTPSProxyURLReturns("https://proxy.com")
 		fakeDBWorker.NoProxyReturns("http://noproxy.com")
-		fakeDBWorker.CreateContainerReturns(fakeCreatingContainer, nil)
 
 		gardenWorker = NewGardenWorker(
 			fakeGardenClient,
@@ -1520,16 +1519,46 @@ var _ = Describe("Worker", func() {
 		})
 
 		Context("when container does not exist in database", func() {
+
 			BeforeEach(func() {
-				fakeDBWorker.CreateContainerReturns(fakeCreatingContainer, nil)
 				fakeDBWorker.FindContainerReturns(nil, nil, nil)
+				fakeDBWorker.CreateContainerReturns(fakeCreatingContainer, nil)
 			})
 
-			It("creates a creating container in database", func() {
+			It("attemps to create container in the db", func() {
 				Expect(fakeDBWorker.CreateContainerCallCount()).To(Equal(1))
-				owner, metadata := fakeDBWorker.CreateContainerArgsForCall(0)
-				Expect(owner).To(Equal(fakeContainerOwner))
-				Expect(metadata).To(Equal(containerMetadata))
+			})
+
+			Context("having db container creation erroring", func() {
+				Context("with ContainerOwnerDisappearedError", func() {
+					BeforeEach(func() {
+						fakeDBWorker.CreateContainerReturns(nil, db.ContainerOwnerDisappearedError{})
+					})
+
+					It("fails w/ ResourceConfigCheckSessionExpiredError", func() {
+						Expect(findOrCreateErr).To(HaveOccurred())
+						Expect(findOrCreateErr).To(Equal(ResourceConfigCheckSessionExpiredError))
+					})
+				})
+
+				Context("with a non-specific error", func() {
+					BeforeEach(func() {
+						fakeDBWorker.CreateContainerReturns(nil, errors.New("err"))
+					})
+
+					It("fails with the same err", func() {
+						Expect(findOrCreateErr).To(HaveOccurred())
+						Expect(findOrCreateErr).To(Equal(errors.New("err")))
+					})
+				})
+			})
+
+			Context("having db container creation succeeding", func() {
+				It("creates a creating container in database", func() {
+					owner, metadata := fakeDBWorker.CreateContainerArgsForCall(0)
+					Expect(owner).To(Equal(fakeContainerOwner))
+					Expect(metadata).To(Equal(containerMetadata))
+				})
 			})
 
 		})

--- a/release-notes/v5.8.0.md
+++ b/release-notes/v5.8.0.md
@@ -131,3 +131,7 @@
 #### <sub><sup><a name="4895" href="#4895">:link:</a></sup></sub> fix
 
 * @kirillbilchenko fixed a [bug](https://github.com/concourse/concourse/issues/3856) where the `concourse_workers_registered` metric would never go below 1, even when workers were pruned. #4895
+
+#### <sub><sup><a name="4947" href="#4947">:link:</a></sup></sub> fix
+
+* @cirocosta fixed a [bug](https://github.com/concourse/concourse/issues/4932) where an error that's not specific could lead to null pointer exception during the container creation phase. #4932

